Neos.EventSourcedNeosAdjustments/Classes/Ui/Controller/ContentController.php

```diff
@@ -0,0 +1,433 @@
+                    continue;
+                }
+                $q = new FlowQuery([$pluginViewNode]);
+                $page = $q->closest('[instanceof Neos.Neos:Document]')->get(0);
+                $uri = $this->uriBuilder
+                    ->reset()
+                    ->uriFor('show', ['node' => $page], 'Frontend\Node', 'Neos.Neos');
+                $views[$pluginViewDefinition->getName()] = [
+                    'label' => $label,
+                    'pageNode' => [
+                        'title' => $page->getLabel(),
+                        'uri' => $uri
+                    ]
+                ];
+            }
+        }
+        return json_encode((object) $views);
+    }
+
+    /**
+     * Fetch all master plugins that are available in the current
+     * workspace.
+     *
+     * @param string $workspaceName Name of the workspace to use for querying the node
+     * @param array $dimensions Optional list of dimensions and their values which should be used for querying the specified node
+     * @return string JSON encoded array of node path => label
+     * @throws \Neos\Eel\Exception
+     */
+    public function masterPluginsAction($workspaceName = 'live', array $dimensions = [])
+    {
+        $this->response->setContentType('application/json');
+
+        $contentContext = $this->createContentContext($workspaceName, $dimensions);
+        $pluginNodes = $this->pluginService->getPluginNodesWithViewDefinitions($contentContext);
+
+        $masterPlugins = [];
+        if (is_array($pluginNodes)) {
+            /** @var $pluginNode NodeInterface */
+            foreach ($pluginNodes as $pluginNode) {
+                if ($pluginNode->isRemoved()) {
+                    continue;
+                }
+                $q = new FlowQuery([$pluginNode]);
+                $page = $q->closest('[instanceof Neos.Neos:Document]')->get(0);
+                if ($page === null) {
+                    continue;
+                }
+                $translationHelper = new TranslationHelper();
+                $masterPlugins[$pluginNode->getIdentifier()] = $translationHelper->translate(
+                    'masterPlugins.nodeTypeOnPageLabel',
+                    null,
+                    ['nodeTypeName' => $translationHelper->translate($pluginNode->getNodeType()->getLabel()), 'pageLabel' => $page->getLabel()],
+                    'Main',
+                    'Neos.Neos'
+                );
+            }
+        }
+        return json_encode((object) $masterPlugins);
+    }
+
+    /**
+     * Signals that a new asset has been uploaded through the Neos Backend
+     *
+     * @param Asset $asset The uploaded asset
+     * @param NodeAddress $node The node the asset belongs to
+     * @param string $propertyName The node property name the asset is assigned to
+     * @return void
+     * @Flow\Signal
+     */
+    protected function emitAssetUploaded(Asset $asset, NodeAddress $node, string $propertyName)
```

Comment:
short question, doesn't this emit a **new** signal compared to the original one; because the class name is different? How can it be that then the original slot is triggered? (I don't see any Package.php code)

Reply:
yes, I guess so – I'll check this.

And, as we just discussed, test if we can provide this adjustment as a Composer patch rather than an aspect.